src/main/java/org/opensearch/flowframework/transport/CreateWorkflowTransportAction.java

```diff
@@ -8,63 +8,104 @@
+            request.getTemplate().templateVersion(),
+            request.getTemplate().compatibilityVersion(),
+            request.getTemplate().workflows(),
+            user
+        );
         if (request.getWorkflowId() == null) {
             // Create new global context and state index entries
-            globalContextHandler.putTemplateToGlobalContext(request.getTemplate(), ActionListener.wrap(response -> {
-                // TODO : Check if state index exists, create if not
-                // TODO : Create StateIndexRequest for workflowId, default to NOT_STARTED
-                listener.onResponse(new WorkflowResponse(response.getId()));
+            flowFrameworkIndicesHandler.putTemplateToGlobalContext(templateWithUser, ActionListener.wrap(globalContextResponse -> {
+                flowFrameworkIndicesHandler.putInitialStateToWorkflowState(
+                    globalContextResponse.getId(),
+                    user,
+                    ActionListener.wrap(stateResponse -> {
+                        logger.info("create state workflow doc " + stateResponse);
+                        listener.onResponse(new WorkflowResponse(globalContextResponse.getId()));
+                    }, exception -> {
+                        logger.error("Failed to save workflow state : {}", exception.getMessage());
+                        listener.onFailure(new FlowFrameworkException(exception.getMessage(), RestStatus.BAD_REQUEST));
+                    })
+                );
             }, exception -> {
                 logger.error("Failed to save use case template : {}", exception.getMessage());
                 listener.onFailure(new FlowFrameworkException(exception.getMessage(), RestStatus.INTERNAL_SERVER_ERROR));
             }));
         } else {
             // Update existing entry, full document replacement
-            globalContextHandler.updateTemplateInGlobalContext(
+            flowFrameworkIndicesHandler.updateTemplateInGlobalContext(
                 request.getWorkflowId(),
                 request.getTemplate(),
                 ActionListener.wrap(response -> {
-                    // TODO : Create StateIndexRequest for workflowId to reset entry to NOT_STARTED
-                    listener.onResponse(new WorkflowResponse(response.getId()));
+                    flowFrameworkIndicesHandler.updateWorkflowState(
+                        request.getWorkflowId(),
+                        ImmutableMap.of(STATE_FIELD, State.NOT_STARTED, PROVISIONING_PROGRESS_FIELD, ProvisioningProgress.NOT_STARTED),
+                        ActionListener.wrap(updateResponse -> {
+                            logger.info("updated workflow {} state to NOT_STARTED", request.getWorkflowId());
+                            listener.onResponse(new WorkflowResponse(request.getWorkflowId()));
+                        }, exception -> {
+                            logger.error("Failed to update workflow state : {}", exception.getMessage());
```

Comment:
What will be state here when the update failed?

Reply:
In this case we are updating an existing workflow, if we fail to update the state doc then the state doc here will be the same as it used to be. The information regarding provisioning time and more could be wrong here. In AD when we update a detector for example we need to stop the task beforehand anyways for example so the task is anyways gone before we update. Maybe we should do something similar for us too where we stop the workflow before updating and then that means we either reset or state document there or we delete it completely. When we update and restart the workflow we can create a new state document.

We might need a little longer discussion on updating in general.

Reply:
Created an issue for initial problem discussion, dont have a solution here yet https://github.com/opensearch-project/opensearch-ai-flow-framework/issues/122